### DISCUSSION TITLE
get tetrahedron coordinates from triangulation

### DISCUSSION THREAD

**Discussion by oren-pinkas** [pyvista/pyvista#3144] [Q&A] [answered]
Greetings!
My problem is similar to #1465, but with tetrahedrons.
I send a list of 3d coordinates to delaunay_3d() and I obtain a 3d triangulation.
I would like to be able to extract the coordinates of the 4 vertices constructing each tetrahedron, so something of the shape (m, 4, 3)
for m being the number of tetrahedrons created by delaunay_3d(), or (m, 4) only for indices.

Can you help me please?
Thank you very much!

**Comment by MatthewFlamm** [comment, marked as answer]
Here is an example to get the information for a single cell.  Then you can loop over all the cells.

```py
import pyvista as pv
from pyvista import examples
mesh = examples.load_tetbeam()  # UnstructuredGrid
# print information about cell 0
print(mesh.cell_type(0))  # 10 is Tetra
print(mesh.cell_points(0))  # (4, 3) points for cell 0
print(mesh.cell_point_ids(0))  # (4,) point ids for cell 0
```

Yields:

```
10
[[0.  0.  0. ]
 [0.5 0.  0. ]
 [0.  0.  0.5]
 [0.  0.5 0. ]]
[0, 1, 9, 3]
```

**Reply by oren-pinkas** [reply]
Thank you!

**Comment by adeak** [comment]
To add to @MatthewFlamm's answer, you can also use a bit of NumPy do get the `(n_cells, 4, 3)` shaped array you need. Since after a tetrahedralization we know that every cell is a tetrahedron, we can reshape the `cells` attribute to grab the `(n_cells, 4)`-shaped index array that gives us the four points for each tetrahedron. Then we can use these to index into the `points` array.
```py
import pyvista as pv
from pyvista import examples
mesh = examples.load_tetbeam()

# check that everything is a tetrahedron
assert (mesh.cells[::5] == 10).all()
# reshape point indices, throw away first column of all "10"
inds = mesh.cells.reshape(-1, 5)[:, 1:]  # shape (n_cells, 4) ints
# use 2d indexing into the points array
tetra_points = mesh.points[inds]  # shape (n_cells, 4, 3) floats

# confirm with a "random" cell
import numpy as np
cell_index = 42
assert np.allclose(tetra_points[cell_index, ...], mesh.cell_points(cell_index))
```

Actually, this is the exact same solution as what was posted for discussion https://github.com/pyvista/pyvista/discussions/1465 you linked :) It doesn't matter what the type of cells are as long as they all have the same number of points in a cell.
